### PR TITLE
KVStore: Distinguish different versions of Proxy

### LINKED ISSUE #8382: Introduce two stages FAP

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What problem does this PR solve?

Issue Number: close #8382

Problem Summary:

### What is changed and how it works?

```commit-message
Distinguish different versions of Proxy
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by yongman — APPROVED** [review]

File: cmake/find_tiflash_proxy.cmake [hunk 1]
```diff
@@ -48,6 +48,9 @@ if(NOT EXTERNAL_TIFLASH_PROXY_FOUND)
 endif()
 
 set(TIFLASH_PROXY_FOUND TRUE)
+# SERVERLESS_PROXY=0 if using normal proxy.
+# SERVERLESS_PROXY=1 if using serverless proxy.
+add_definitions(-DSERVERLESS_PROXY=0)
```
Comment (Lloyd-Pottiger) [inline cmake/find_tiflash_proxy.cmake:53]: How about

```cmake
if (EXISTS "${TiFlash_SOURCE_DIR}/contrib/tiflash-proxy/xxx/xxx")
   add_definitions(-DSERVERLESS_PROXY=1)
else ()
   add_definitions(-DSERVERLESS_PROXY=0)
endif ()
```

Comment (CalvinNeo) [reply cmake/find_tiflash_proxy.cmake:53]: OK, I can use rfstore

File: dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeer.cpp [hunk 2]
```diff
@@ -328,8 +328,12 @@ FastAddPeerRes FastAddPeerImplWrite(
         GET_METRIC(tiflash_fap_task_result, type_failed_cancel).Increment();
         return genFastAddPeerRes(FastAddPeerStatus::Canceled, "", "");
     }
+
+    Stopwatch sbuild;
     auto segments = dm_storage->buildSegmentsFromCheckpointInfo(new_key_range, checkpoint_info, settings);
+    GET_METRIC(tiflash_fap_task_duration_seconds, type_write_stage_build).Observe(sbuild.elapsedSeconds());
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeer.cpp:334]: use `elapsedSecondsFromLastTime()` can avoid build multiple watch

File: dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp [hunk 3]
```diff
@@ -323,8 +323,11 @@ static inline std::pair<std::vector<std::string>, size_t> getSplitKey(
 
     // Don't change the order of following checks, `getApproxBytes` involves some overhead,
     // although it is optimized to bring about the minimum overhead.
-    if (new_region->getClusterRaftstoreVer() != RaftstoreVer::V2)
-        return std::make_pair(std::vector<std::string>{}, 0);
+    if constexpr (SERVERLESS_PROXY == 0)
+    {
+        if (new_region->getClusterRaftstoreVer() != RaftstoreVer::V2)
+            return std::make_pair(std::vector<std::string>{}, 0);
+    }
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp:330]:
```suggestion
#if SERVERLESS_PROXY == 0
        if (new_region->getClusterRaftstoreVer() != RaftstoreVer::V2)
            return std::make_pair(std::vector<std::string>{}, 0);
#endif
```

File: dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeer.cpp (hunk 2 above)
Comment (CalvinNeo) [reply dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeer.cpp]: fixed

**Review by Lloyd-Pottiger — APPROVED** [review]

**Pull request merged**